### ISSUE TITLE
BUG: Cannot read GDB file using a geometry filter (mask)

### ISSUE THREAD

**Issue by shiyuning** [geopandas/geopandas#3444]
- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of geopandas.

- [ ] (optional) I have confirmed this bug exists on the main branch of geopandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
gdf = gpd.read_file(
    example.gdb,
    driver='FileGDB',
    layer='example_layer',
    mask=mask_gdf,
```

#### Problem description

When running the above code using Geopandas v1.0.1, I got the following error: `ValueError: Must pass either crs or epsg.`.
The code above, however, works with v0.14.4

#### Output of ``geopandas.show_versions()``

<details>

SYSTEM INFO
-----------
python     : 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]
executable : /usr/bin/python3
machine    : Linux-6.1.85+-x86_64-with-glibc2.35

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.11.4
GEOS lib   : None
GDAL       : 3.9.1
GDAL data dir: /usr/local/lib/python3.10/dist-packages/pyogrio/gdal_data/
PROJ       : 9.4.1
PROJ data dir: /usr/local/lib/python3.10/dist-packages/pyproj/proj_dir/share/proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 1.0.1
numpy      : 1.26.4
pandas     : 2.2.2
pyproj     : 3.7.0
shapely    : 2.0.6
pyogrio    : 0.10.0
geoalchemy2: None
geopy      : 2.4.1
matplotlib : 3.7.1
mapclassify: None
fiona      : 1.10.1
psycopg    : None
psycopg2   : 2.9.9 (dt dec pq3 ext lo64)
pyarrow    : 16.1.0

</details>

**Comment by martinfleis** [comment]
Can you show the full traceback (error message) please?

**Comment by shiyuning** [comment]
Here it is:

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-bb0d5cb569b5> in <cell line: 11>()
      9 
     10 print(boundary)
---> 11 _soil = gpd.read_file(
     12     GSSURGO(state),
     13     driver='FileGDB',

5 frames
/usr/local/lib/python3.10/dist-packages/geopandas/io/file.py in _read_file(filename, bbox, mask, columns, rows, engine, **kwargs)
    292 
    293     if engine == "pyogrio":
--> 294         return _read_file_pyogrio(
    295             filename, bbox=bbox, mask=mask, columns=columns, rows=rows, **kwargs
    296         )

/usr/local/lib/python3.10/dist-packages/geopandas/io/file.py in _read_file_pyogrio(path_or_bytes, bbox, mask, rows, **kwargs)
    499                 path_or_bytes.seek(0)
    500 
--> 501             mask = shapely.unary_union(mask.to_crs(crs).geometry.values)
    502         elif isinstance(mask, BaseGeometry):
    503             mask = shapely.unary_union(mask)

/usr/local/lib/python3.10/dist-packages/geopandas/geodataframe.py in to_crs(self, crs, epsg, inplace)
   1699         else:
   1700             df = self.copy()
-> 1701         geom = df.geometry.to_crs(crs=crs, epsg=epsg)
   1702         df.geometry = geom
   1703         if not inplace:

/usr/local/lib/python3.10/dist-packages/geopandas/geoseries.py in to_crs(self, crs, epsg)
   1205         """
   1206         return GeoSeries(
-> 1207             self.values.to_crs(crs=crs, epsg=epsg), index=self.index, name=self.name
   1208         )
   1209 

/usr/local/lib/python3.10/dist-packages/geopandas/_compat.py in wrapper(*args, **kwargs)
     88                 f"\nImporting pyproj resulted in: {pyproj_import_error}"
     89             )
---> 90         return func(*args, **kwargs)
     91 
     92     return wrapper

/usr/local/lib/python3.10/dist-packages/geopandas/array.py in to_crs(self, crs, epsg)
   1009             crs = CRS.from_epsg(epsg)
   1010         else:
-> 1011             raise ValueError("Must pass either crs or epsg.")
   1012 
   1013         # skip if the input CRS and output CRS are the exact same

ValueError: Must pass either crs or epsg.

**Comment by brendan-ward** [comment]
This error happens when the source dataset does not have a CRS defined and you pass a mask as a GeoDataFrame, because we read the CRS from the source dataset and attempt to transform the mask to match it.

Can you confirm that the source file does not have a CRS defined?

```Python
from pyogrio import read_info

print(read_info(example.gdb).get('crs'))
# should be None for above errror
```

While we can check if the CRS is `None` and then not try to match the mask to it (provided the CRS of the mask is also `None`), that's about all we can do here; if the mask has a CRS but the source dataset does not, we'd have to (or should) raise an error about not being able to align them.

In this case, it looks like the mask GeoDataFrame has a CRS defined, or you'd get a different error.

To avoid this error, just pass in the mask as a Shapely geometry:

```Python
import shapely

gdf = gpd.read_file(
    example.gdb,
    driver='FileGDB',
    layer='example_layer',
    mask=shapely.union_all(mask_gdf.geometry.values),
```

Note that this only works if both the mask and source dataset are in the same coordinate system, or lack thereof; otherwise their coordinates will not necessarily line up.

**Comment by shiyuning** [comment]
Yes, @brendan-ward , you are absolutely right. The source file does not have a CRS defined, and  your suggestion worked perfectly. Thank you very much for your help! I will close the issue since it is resolved.

**Comment by jorisvandenbossche** [comment]
Reopening this, because I think we can still provide an error message to essentially say to the user what has been discussed above ("specified mask has a CRS but the source dataset has not, pass the mask GeoDataFrame without a CRS as well or as a shapely Polygon")

**Comment by martinfleis** [comment]
@iamabhinav-dev this is already being resolved in #3464

**Comment by RemoteCMeijer** [comment]
First time contributor to the project, so I might be wrong, but I see that #3464 has been merged. Should this issue not be closed?
